### PR TITLE
Updated shadowing logic to produce valid OSGI MANIFEST.MF

### PR DESCRIPTION
**Pull request by rozza**
JAVA-2455

### CODE REVIEW

**Comment by rozza** [comment]
bson MANIFEST.MF

```
Manifest-Version: 1.0
Export-Package: org.bson;version="3.5.0.SNAPSHOT";uses:="org.bson.code
 cs,org.bson.codecs.configuration,org.bson.conversions,org.bson.io,org
 .bson.json,org.bson.types",org.bson.assertions;version="3.5.0.SNAPSHO
 T",org.bson.codecs;version="3.5.0.SNAPSHOT";uses:="org.bson,org.bson.
 codecs.configuration,org.bson.types",org.bson.codecs.configuration;ve
 rsion="3.5.0.SNAPSHOT";uses:="org.bson.codecs",org.bson.codecs.pojo;v
 ersion="3.5.0.SNAPSHOT";uses:="org.bson.codecs,org.bson.codecs.config
 uration",org.bson.codecs.pojo.annotations;version="3.5.0.SNAPSHOT",or
 g.bson.conversions;version="3.5.0.SNAPSHOT";uses:="org.bson,org.bson.
 codecs.configuration",org.bson.diagnostics;version="3.5.0.SNAPSHOT",o
 rg.bson.internal;version="3.5.0.SNAPSHOT",org.bson.io;version="3.5.0.
 SNAPSHOT";uses:="org.bson,org.bson.types",org.bson.json;version="3.5.
 0.SNAPSHOT";uses:="org.bson,org.bson.types",org.bson.shaded.fasterxml
 .classmate;version="3.5.0.SNAPSHOT";uses:="org.bson.shaded.fasterxml.
 classmate.members,org.bson.shaded.fasterxml.classmate.types,org.bson.
 shaded.fasterxml.classmate.util",org.bson.shaded.fasterxml.classmate.
 members;version="3.5.0.SNAPSHOT";uses:="org.bson.shaded.fasterxml.cla
 ssmate,org.bson.shaded.fasterxml.classmate.util",org.bson.shaded.fast
 erxml.classmate.types;version="3.5.0.SNAPSHOT";uses:="org.bson.shaded
 .fasterxml.classmate,org.bson.shaded.fasterxml.classmate.members",org
 .bson.shaded.fasterxml.classmate.util;version="3.5.0.SNAPSHOT";uses:=
 "org.bson.shaded.fasterxml.classmate,org.bson.shaded.fasterxml.classm
 ate.types",org.bson.types;version="3.5.0.SNAPSHOT";uses:="org.bson"
Bundle-SymbolicName: org.mongodb.bson
Bundle-Version: 3.5.0.SNAPSHOT
Bundle-Name: bson
Bundle-ManifestVersion: 2
Bnd-LastModified: 1491305490000
Import-Package: javax.xml.bind
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.6))"
Created-By: 1.8.0_31 (Oracle Corporation)
Tool: Bnd-2.4.0.201411031534
Build-Version: 3.4.0-rc1-105-g2a2b3c2-dirty
```

mongo-java-driver

```
Manifest-Version: 1.0
Export-Package: com.mongodb;version="3.5.0.SNAPSHOT";uses:="com.mongod
 b.annotations,com.mongodb.bulk,com.mongodb.client,com.mongodb.client.
 model,com.mongodb.connection,com.mongodb.event,javax.net,org.bson,org
 .bson.codecs,org.bson.codecs.configuration,org.bson.conversions,org.b
 son.io,org.bson.json,org.bson.types",com.mongodb.annotations;version=
 "3.5.0.SNAPSHOT",com.mongodb.assertions;version="3.5.0.SNAPSHOT";uses
 :="com.mongodb.async",com.mongodb.async;version="3.5.0.SNAPSHOT",com.
 mongodb.binding;version="3.5.0.SNAPSHOT";uses:="com.mongodb,com.mongo
 db.async,com.mongodb.connection",com.mongodb.bulk;version="3.5.0.SNAP
 SHOT";uses:="com.mongodb,com.mongodb.client.model,org.bson",com.mongo
 db.client;version="3.5.0.SNAPSHOT";uses:="com.mongodb,com.mongodb.ann
 otations,com.mongodb.bulk,com.mongodb.client.model,com.mongodb.client
 .result,org.bson,org.bson.codecs.configuration,org.bson.conversions",
 com.mongodb.client.gridfs;version="3.5.0.SNAPSHOT";uses:="com.mongodb
 ,com.mongodb.annotations,com.mongodb.client,com.mongodb.client.gridfs
 .model,com.mongodb.client.model,org.bson,org.bson.conversions,org.bso
 n.types",com.mongodb.client.gridfs.codecs;version="3.5.0.SNAPSHOT";us
 es:="com.mongodb.client.gridfs.model,org.bson,org.bson.codecs,org.bso
 n.codecs.configuration",com.mongodb.client.gridfs.model;version="3.5.
 0.SNAPSHOT";uses:="org.bson,org.bson.types",com.mongodb.client.jndi;v
 ersion="3.5.0.SNAPSHOT";uses:="javax.naming,javax.naming.spi",com.mon
 godb.client.model;version="3.5.0.SNAPSHOT";uses:="com.mongodb,com.mon
 godb.annotations,com.mongodb.client.model.geojson,org.bson,org.bson.c
 onversions",com.mongodb.client.model.geojson;version="3.5.0.SNAPSHOT"
 ;uses:="com.mongodb.annotations",com.mongodb.client.model.geojson.cod
 ecs;version="3.5.0.SNAPSHOT";uses:="com.mongodb.client.model.geojson,
 org.bson,org.bson.codecs,org.bson.codecs.configuration",com.mongodb.c
 lient.result;version="3.5.0.SNAPSHOT";uses:="org.bson",com.mongodb.co
 nnection;version="3.5.0.SNAPSHOT";uses:="com.mongodb,com.mongodb.anno
 tations,com.mongodb.async,com.mongodb.binding,com.mongodb.bulk,com.mo
 ngodb.client,com.mongodb.event,com.mongodb.internal.connection,com.mo
 ngodb.selector,javax.net,org.bson,org.bson.codecs,org.bson.io,org.bso
 n.types",com.mongodb.connection.netty;version="3.5.0.SNAPSHOT";uses:=
 "com.mongodb,com.mongodb.connection,io.netty.buffer,io.netty.channel,
 io.netty.channel.socket",com.mongodb.diagnostics.logging;version="3.5
 .0.SNAPSHOT",com.mongodb.event;version="3.5.0.SNAPSHOT";uses:="com.mo
 ngodb.annotations,com.mongodb.connection,org.bson",com.mongodb.gridfs
 ;version="3.5.0.SNAPSHOT";uses:="com.mongodb,org.bson,org.bson.types"
 ,com.mongodb.internal;version="3.5.0.SNAPSHOT",com.mongodb.internal.a
 sync;version="3.5.0.SNAPSHOT";uses:="com.mongodb.async,com.mongodb.di
 agnostics.logging",com.mongodb.internal.authentication;version="3.5.0
 .SNAPSHOT";uses:="org.bson",com.mongodb.internal.connection;version="
 3.5.0.SNAPSHOT";uses:="com.mongodb,com.mongodb.connection,javax.net.s
 sl,org.bson",com.mongodb.internal.management.jmx;version="3.5.0.SNAPS
 HOT";uses:="com.mongodb.management",com.mongodb.internal.thread;versi
 on="3.5.0.SNAPSHOT",com.mongodb.internal.validator;version="3.5.0.SNA
 PSHOT";uses:="org.bson",com.mongodb.management;version="3.5.0.SNAPSHO
 T";uses:="com.mongodb.event",com.mongodb.operation;version="3.5.0.SNA
 PSHOT";uses:="com.mongodb,com.mongodb.annotations,com.mongodb.async,c
 om.mongodb.binding,com.mongodb.bulk,com.mongodb.client.model,com.mong
 odb.connection,org.bson,org.bson.codecs",com.mongodb.selector;version
 ="3.5.0.SNAPSHOT";uses:="com.mongodb,com.mongodb.annotations,com.mong
 odb.connection",com.mongodb.util;version="3.5.0.SNAPSHOT";uses:="org.
 bson",org.bson;version="3.5.0.SNAPSHOT";uses:="org.bson.codecs,org.bs
 on.codecs.configuration,org.bson.conversions,org.bson.io,org.bson.jso
 n,org.bson.types",org.bson.assertions;version="3.5.0.SNAPSHOT",org.bs
 on.codecs;version="3.5.0.SNAPSHOT";uses:="org.bson,org.bson.codecs.co
 nfiguration,org.bson.types",org.bson.codecs.configuration;version="3.
 5.0.SNAPSHOT";uses:="org.bson.codecs",org.bson.codecs.pojo;version="3
 .5.0.SNAPSHOT";uses:="org.bson.codecs,org.bson.codecs.configuration",
 org.bson.codecs.pojo.annotations;version="3.5.0.SNAPSHOT",org.bson.co
 nversions;version="3.5.0.SNAPSHOT";uses:="org.bson,org.bson.codecs.co
 nfiguration",org.bson.diagnostics;version="3.5.0.SNAPSHOT",org.bson.i
 nternal;version="3.5.0.SNAPSHOT",org.bson.io;version="3.5.0.SNAPSHOT"
 ;uses:="org.bson,org.bson.types",org.bson.json;version="3.5.0.SNAPSHO
 T";uses:="org.bson,org.bson.types",org.bson.shaded.fasterxml.classmat
 e;version="3.5.0.SNAPSHOT";uses:="org.bson.shaded.fasterxml.classmate
 .members,org.bson.shaded.fasterxml.classmate.types,org.bson.shaded.fa
 sterxml.classmate.util",org.bson.shaded.fasterxml.classmate.members;v
 ersion="3.5.0.SNAPSHOT";uses:="org.bson.shaded.fasterxml.classmate,or
 g.bson.shaded.fasterxml.classmate.util",org.bson.shaded.fasterxml.cla
 ssmate.types;version="3.5.0.SNAPSHOT";uses:="org.bson.shaded.fasterxm
 l.classmate,org.bson.shaded.fasterxml.classmate.members",org.bson.sha
 ded.fasterxml.classmate.util;version="3.5.0.SNAPSHOT";uses:="org.bson
 .shaded.fasterxml.classmate,org.bson.shaded.fasterxml.classmate.types
 ",org.bson.types;version="3.5.0.SNAPSHOT";uses:="org.bson",org.bson.u
 til;version="3.5.0.SNAPSHOT"
Bundle-SymbolicName: org.mongodb.mongo-java-driver
Bundle-Version: 3.5.0.SNAPSHOT
Bundle-Name: mongo-java-driver
Bundle-ManifestVersion: 2
Bnd-LastModified: 1491305509000
Import-Package: io.netty.bootstrap;version="[4.1,5)",io.netty.buffer;v
 ersion="[4.1,5)",io.netty.channel;version="[4.1,5)",io.netty.channel.
 nio;version="[4.1,5)",io.netty.channel.socket;version="[4.1,5)",io.ne
 tty.channel.socket.nio;version="[4.1,5)",io.netty.handler.ssl;version
 ="[4.1,5)",io.netty.handler.timeout;version="[4.1,5)",io.netty.util.c
 oncurrent;version="[4.1,5)",javax.crypto,javax.crypto.spec,javax.mana
 gement,javax.naming,javax.naming.spi,javax.net,javax.net.ssl,javax.se
 curity.auth,javax.security.auth.callback,javax.security.sasl,javax.xm
 l.bind,org.ietf.jgss,org.slf4j;version="[1.7,2)"
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.6))"
Created-By: 1.8.0_31 (Oracle Corporation)
Tool: Bnd-2.4.0.201411031534
Build-Version: 3.4.0-rc1-105-g2a2b3c2-dirty
```

File: gradle/shadow.gradle [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    configurations {
+        compile.extendsFrom shadow
+    }
+
+    shadowJar {
+        from sourceSets.main.output
+        configurations = [project.configurations.compile]
+        mergeServiceFiles()
+        relocate 'com.fasterxml', 'org.bson.shaded.fasterxml'
```
Comment (jyemin) [inline gradle/shadow.gradle]: Despite this line, shadowJar adds this to the manifest:

    Class-Path: classmate-1.3.1.jar

I don't know if it matters, but it seems incorrect given that we're shading this library

I also see a META-INF/maven/com.fasterxml/classmate directory, containing the classmate pom file, in the bson and mongo-java-driver jar files, and I'm not clear whether it needs to be there.

There's also a NOTICE and LICENSE file from classmate next to our MANIFEST.MF.  If there's going to be a license there it should be ours.

File: gradle/shadow.gradle [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    apply plugin: 'com.github.johnrengelman.shadow'
+
+    configurations {
+        compile.extendsFrom shadow
+    }
+
+    shadowJar {
+        from sourceSets.main.output
+        configurations = [project.configurations.compile]
+        mergeServiceFiles()
```
Comment (jyemin) [inline gradle/shadow.gradle]: Is this line necessary since we have no META-INF/services directory?

File: gradle/shadow.gradle [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        relocate 'com.fasterxml', 'org.bson.shaded.fasterxml'
+        classifier = 'shadow'
+        dependencies {
```
Comment (jyemin) [inline gradle/shadow.gradle]: What is this excluding from exactly?  Does it change the generated pom.xml or something about the jar file itself?

File: gradle/shadow.gradle [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            eachEntry {
+                if (OsgiOverride.OVERRIDE_KEYS.contains(it.getKey())) {
+                    it.setValue(it.getMergeValue())
+                } else {
+                    it.setValue(it.getBaseValue())
+                }
+            }
+        }
+
+        classifier = ''
```
Comment (jyemin) [inline gradle/shadow.gradle]: What does an empty classifier signify?

File: mongo-java-driver/build.gradle [hunk 5]
```diff
@@ -40,6 +29,7 @@ dependencies {
     compile "io.netty:netty-buffer:$nettyVersion", optional
     compile "io.netty:netty-transport:$nettyVersion", optional
     compile "io.netty:netty-handler:$nettyVersion", optional
+    compile project(path: ':driver')
```
Comment (jyemin) [inline mongo-java-driver/build.gradle]: This is necessary to pull in the shaded classmate packages?  How does it interact with the srcDirs on line 49?

File: gradle/shadow.gradle (hunk 1 above)
Comment (jyemin) [inline gradle/shadow.gradle]: WIth the latest commit I'm now seeing a META-INF directory containing just the MANIFEST.MF, which no longer has a Class-Path entry.  So all is well.

File: gradle/shadow.gradle (hunk 2 above)
Comment (jyemin) [inline gradle/shadow.gradle]: Looks like you killed this line in the last patch.

File: gradle/shadow.gradle [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    shadowJar {
+        relocate 'com.fasterxml', 'org.bson.shaded.fasterxml'
+        classifier = 'shadow'
+        dependencies {
+            exclude(dependency('io.netty:.*'))
+            exclude(dependency('org.slf4j:.*'))
+        }
+        append "$rootDir/LICENSE.txt"
```
Comment (jyemin) [inline gradle/shadow.gradle]: I'm not seeing the LICENSE.txt file in the jar file

File: gradle/shadow.gradle (hunk 3 above)
Comment (jyemin) [inline gradle/shadow.gradle]: Wouldn't it only bring dependencies of type 'shadow' into the jar?

**Pull request closed without merging**